internal/attestation/aws/snp/validator.go

```diff
@@ -9,99 +9,211 @@ package snp
+	httpsGetter trust.HTTPSGetter
+}
+
+type reportVerifier interface {
+	SnpAttestation(att *sevsnp.Attestation, opts *verify.Options) error
+}
+type reportValidator interface {
+	SnpAttestation(att *sevsnp.Attestation, opts *validate.Options) error
+}
+
+type reportValidatorImpl struct{}
+
+func (r *reportValidatorImpl) SnpAttestation(att *sevsnp.Attestation, opts *validate.Options) error {
+	return validate.SnpAttestation(att, opts)
+}
+
+type reportVerifierImpl struct{}
+
+func (r *reportVerifierImpl) SnpAttestation(att *sevsnp.Attestation, opts *verify.Options) error {
+	return verify.SnpAttestation(att, opts)
+}
+
+// validate the report by checking if it has a valid VLEK signature.
+// The certificate chain ARK -> ASK -> VLEK is also validated.
+// Checks that the report's userData matches the connection's userData.
+func (a *awsValidator) validate(attestation vtpm.AttestationDocument, ask *x509.Certificate, ark *x509.Certificate, akDigest [64]byte, config *config.AWSSEVSNP, log attestation.Logger) error {
+	var info snp.InstanceInfo
+	if err := json.Unmarshal(attestation.InstanceInfo, &info); err != nil {
+		return newValidationError(fmt.Errorf("unmarshalling instance info: %w", err))
 	}
 
-	imageID := idDocument.ImageID
+	certchain := snp.NewCertificateChain(ask, ark)
 
-	client, err := v.getDescribeClient(ctx, idDocument.Region)
+	att, err := info.AttestationWithCerts(a.httpsGetter, certchain, log)
 	if err != nil {
-		return err
+		return newValidationError(fmt.Errorf("getting attestation with certs: %w", err))
 	}
-	// Currently, there seems to be a problem with retrieving image attributes directly.
-	// Alternatively, parse it from the general output.
-	imageOutput, err := client.DescribeImages(ctx, &ec2.DescribeImagesInput{ImageIds: []string{imageID}})
+
+	verifyOpts, err := getVerifyOpts(att)
 	if err != nil {
-		return err
+		return newValidationError(fmt.Errorf("getting verify options: %w", err))
 	}
-	if len(imageOutput.Images) == 0 {
-		return fmt.Errorf("aws image %s not found", imageID)
+
+	if err := a.verifier.SnpAttestation(att, verifyOpts); err != nil {
+		return newValidationError(fmt.Errorf("verifying SNP attestation: %w", err))
 	}
-	if len(imageOutput.Images) > 1 {
-		return fmt.Errorf("found multiple image references for image ID %s", imageID)
+
+	validateOpts := &validate.Options{
+		// Check that the attestation key's digest is included in the report.
+		ReportData: akDigest[:],
+		GuestPolicy: abi.SnpPolicy{
+			Debug: false, // Debug means the VM can be decrypted by the host for debugging purposes and thus is not allowed.
+			SMT:   true,  // Allow Simultaneous Multi-Threading (SMT). Normally, we would want to disable SMT
+			// but AWS machines are currently facing issues if it's disabled.
+		},
+		VMPL: new(int), // Checks that Virtual Machine Privilege Level (VMPL) is 0.
+		// This checks that the reported LaunchTCB version is equal or greater than the minimum specified in the config.
```

Comment:
We could possibly leave a sentence here explaining why we only check the Launch TCB